crowdsourcehinter/crowdsourcehinter.py

```diff
@@ -0,0 +1,352 @@
+        answer = str(data["submittedanswer"])
+        # put the student's answer to lower case so that differences in capitalization don't make
+        # different groups of hints. this is sloppy and the execution should probably be changed.
+        answer = answer.lower()
+        found_equal_sign = 0
+        remaining_hints = int(0)
+        best_hint = ""
+        # the string returned by the event problem_graded is very messy and is different
+        # for each problem, but after all of the numbers/letters there is an equal sign, after which the
+        # student's input is shown. I use the function below to remove everything before the first equal
+        # sign and only take the student's actual input.
+        #
+        # TODO: figure out better way to directly get text of student's answer
+        if "=" in answer:
+            if found_equal_sign == 0:
+                found_equal_sign = 1
+                eqplace = answer.index("=") + 1
+                answer = answer[eqplace:]
+        remaining_hints = str(self.find_hints(answer))
+        if remaining_hints != str(0):
+            for hint in self.hint_database[str(answer)]:
+                if hint not in self.reported_hints.keys():
+                    # if best_hint hasn't been set yet or the rating of hints is greater than the rating of best_hint
+                    if best_hint == "" or self.hint_database[str(answer)][hint] > self.hint_database[str(answer)][str(best_hint)]:
+                        best_hint = hint
+            self.used.append(best_hint)
+            return {'BestHint': best_hint, "StudentAnswer": answer}
+        # find generic hints for the student if no specific hints exist
+        if len(self.generic_hints) != 0:
+            generic_hint = random.choice(self.generic_hints)
+            self.used.append(generic_hint)
+            return {'BestHint': generic_hint, "StudentAnswer": answer}
+        else:
+            # if there are no hints in either the database or generic hints
+            self.used.append(str("There are no hints for" + " " + answer))
+            return {'BestHint': "Sorry, there are no hints for this answer.", "StudentAnswer": answer}
+
+    def find_hints(self, answer):
+        """
+        This function is used to check that an incorrect answer has available hints to show.
+        It will also add the incorrect answer test to self.incorrect_answers.
+        Args:
+          answer: This is equal to answer from get_hint, the answer the student submitted
+        Returns 0 if no hints to show exist
+        """
+        isreported = []
+        self.incorrect_answers.append(str(answer))
+        if str(answer) not in self.hint_database:
+            # add incorrect answer to hint_database if no precedent exists
+            self.hint_database[str(answer)] = {}
+            return str(0)
+        for hint_keys in self.hint_database[str(answer)]:
```

Comment:
Minor: "hint_keys" is a confusing variable name here, shouldn't it be "hint_key"?